### PR TITLE
Add unit test class for StorageImpl

### PR DESCRIPTION
**Pull request by mziccard**
This again goes in the direction of #14.
It adds a test class for `StorageImpl`. `StorageRpc` and `StorageOptions` are mocked.
Comments are welcome, as usual.

### CODE REVIEW

File: gcloud-java-storage/src/test/java/com/google/gcloud/storage/StorageImplTest.java [hunk 1]
```diff
@@ -0,0 +1,830 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.gcloud.storage;
+
+import static com.google.common.base.Charsets.UTF_8;
```
Comment (aozarov) [inline gcloud-java-storage/src/test/java/com/google/gcloud/storage/StorageImplTest.java]: We should use [this](http://docs.oracle.com/javase/7/docs/api/java/nio/charset/StandardCharsets.html#UTF_8) instead.

File: gcloud-java-storage/src/test/java/com/google/gcloud/storage/StorageImplTest.java [hunk 2]
```diff
@@ -0,0 +1,830 @@
...
+
+  @Before
+  public void setUp() throws IOException, InterruptedException {
+    optionsMock = EasyMock.createMock(StorageOptions.class);
+    storageRpcMock = EasyMock.createMock(StorageRpc.class);
+  }
+
+  @After
+  public void tearDown() throws Exception {
+    EasyMock.verify(optionsMock);
```
Comment (aozarov) [inline gcloud-java-storage/src/test/java/com/google/gcloud/storage/StorageImplTest.java]: If you prefer, verify accepts varargs.

File: gcloud-java-storage/src/test/java/com/google/gcloud/storage/StorageImplTest.java [hunk 3]
```diff
@@ -0,0 +1,830 @@
...
+    EasyMock.verify(storageRpcMock);
+  }
+
+  @Test
+  public void testGetOptions() {
+    EasyMock.expect(optionsMock.storageRpc()).andReturn(storageRpcMock);
+    EasyMock.replay(optionsMock);
```
Comment (aozarov) [inline gcloud-java-storage/src/test/java/com/google/gcloud/storage/StorageImplTest.java]: If you prefer, reply accepts varags

File: gcloud-java-storage/src/test/java/com/google/gcloud/storage/StorageImplTest.java [hunk 4]
```diff
@@ -0,0 +1,830 @@
...
+  public void testListBucketsEmpty() {
+    EasyMock.expect(optionsMock.storageRpc()).andReturn(storageRpcMock).times(2);
+    EasyMock.expect(optionsMock.retryParams()).andReturn(RetryParams.noRetries());
+    EasyMock.replay(optionsMock);
+    storage = StorageFactory.instance().get(optionsMock);
+    EasyMock.expect(storageRpcMock.list(EMPTY_RPC_OPTIONS)).andReturn(
+        Tuple.<String, Iterable<com.google.api.services.storage.model.Bucket>>of(null, null));
+    EasyMock.replay(storageRpcMock);
+    ListResult<BucketInfo> listResult = storage.list();
+    assertEquals(null, listResult.nextPageCursor());
```
Comment (aozarov) [inline gcloud-java-storage/src/test/java/com/google/gcloud/storage/StorageImplTest.java]: assertNull ?

File: gcloud-java-storage/src/test/java/com/google/gcloud/storage/StorageImplTest.java [hunk 5]
```diff
@@ -0,0 +1,830 @@
...
+  public void testListBlobsWithOptions() {
+    EasyMock.expect(optionsMock.storageRpc()).andReturn(storageRpcMock).times(2);
+    EasyMock.expect(optionsMock.retryParams()).andReturn(RetryParams.noRetries());
+    EasyMock.replay(optionsMock);
+    storage = StorageFactory.instance().get(optionsMock);
+    String cursor = "cursor";
+    ImmutableList<BlobInfo> blobList = ImmutableList.of(BLOB_INFO1, BLOB_INFO2);
+    Tuple<String, Iterable<com.google.api.services.storage.model.StorageObject>> result =
+        Tuple.of(cursor, Iterables.transform(blobList,
+            new Function<BlobInfo, com.google.api.services.storage.model.StorageObject>() {
```
Comment (aozarov) [inline gcloud-java-storage/src/test/java/com/google/gcloud/storage/StorageImplTest.java]: Here and in other places. Any reason not to use the package scope functions defined in [BlobInfo](https://github.com/GoogleCloudPlatform/gcloud-java/blob/master/gcloud-java-storage/src/main/java/com/google/gcloud/storage/BlobInfo.java#L54).

File: gcloud-java-storage/src/test/java/com/google/gcloud/storage/StorageImplTest.java [hunk 6]
```diff
@@ -0,0 +1,830 @@
...
+    EasyMock.expect(
+        storageRpcMock.copy(BLOB_INFO2.toPb(), BLOB_SOURCE_OPTIONS_COPY, BLOB_INFO1.toPb(),
+            BLOB_TARGET_OPTIONS_COMPOSE)).andReturn(BLOB_INFO1.toPb());
+    EasyMock.replay(storageRpcMock);
+    BlobInfo blob = storage.copy(req);
+    assertEquals(BLOB_INFO1, blob);
+  }
+
+  @Test
+  public void testReadAllBytes() {
```
Comment (aozarov) [inline gcloud-java-storage/src/test/java/com/google/gcloud/storage/StorageImplTest.java]: testReadAllBytes, testReader and testWriter with options?

File: gcloud-java-storage/src/test/java/com/google/gcloud/storage/StorageImplTest.java [hunk 7]
```diff
@@ -0,0 +1,830 @@
...
+        storageRpcMock.load(BlobInfo.of(BUCKET_NAME1, BLOB_NAME1).toPb(), EMPTY_RPC_OPTIONS))
+        .andReturn(BLOB_CONTENT);
+    EasyMock.replay(storageRpcMock);
+    byte[] readBytes = storage.readAllBytes(BUCKET_NAME1, BLOB_NAME1);
+    assertArrayEquals(BLOB_CONTENT, readBytes);
+  }
+
+  @Test
+  public void testApply() {
+    BatchRequest req =
```
Comment (aozarov) [inline gcloud-java-storage/src/test/java/com/google/gcloud/storage/StorageImplTest.java]: I think this warping is nicer in this case:
```java
BatchRequest req =  BatchRequest.builder()
    .delete(BUCKET_NAME1, BLOB_NAME1)
    .update(BLOB_INFO2)
    .get(BUCKET_NAME1, BLOB_NAME3)
    .build();
```

Are you using any of the formatters provided [here](https://github.com/google/styleguide)?

File: gcloud-java-storage/src/test/java/com/google/gcloud/storage/StorageImplTest.java [hunk 8]
```diff
@@ -0,0 +1,830 @@
...
+    BlobWriteChannel channel = storage.writer(BLOB_INFO1);
+    assertNotNull(channel);
+    assertTrue(channel.isOpen());
+  }
+
+  @Test
+  public void testSignUrl() throws NoSuchAlgorithmException, InvalidKeyException,
+      SignatureException, UnsupportedEncodingException {
+    String account = "account";
+    KeyPair keyPair = KeyPairGenerator.getInstance("RSA").generateKeyPair();
```
Comment (aozarov) [inline gcloud-java-storage/src/test/java/com/google/gcloud/storage/StorageImplTest.java]: Is that fast enough?

**Comment by aozarov** [comment]
Thanks @mziccard. Great work!

I think we should also have a test for exception handling (retryable and non-retryable) similar to the ones
we have for datastore (though could be done in a separate PR).

File: gcloud-java-storage/src/test/java/com/google/gcloud/storage/StorageImplTest.java (hunk 5 above)
Comment (mziccard) [inline gcloud-java-storage/src/test/java/com/google/gcloud/storage/StorageImplTest.java]: @aozarov No reason actually, I forgot about them:)

File: gcloud-java-storage/src/test/java/com/google/gcloud/storage/StorageImplTest.java (hunk 7 above)
Comment (mziccard) [inline gcloud-java-storage/src/test/java/com/google/gcloud/storage/StorageImplTest.java]: @aozarov Actually yes, so I am surprised we are have all these formatting problems with my PRs. I'll double check.

File: gcloud-java-storage/src/test/java/com/google/gcloud/storage/StorageImplTest.java (hunk 8 above)
Comment (mziccard) [inline gcloud-java-storage/src/test/java/com/google/gcloud/storage/StorageImplTest.java]: @aozarov It does not seem to be taking too long to me. By default it should generate a key of length 1024 which should take between 1 and 10 seconds on a relatively recent machine. If you prefer we can always embed private and public keys as strings constants.

File: gcloud-java-storage/src/test/java/com/google/gcloud/storage/StorageImplTest.java (hunk 6 above)
Comment (mziccard) [inline gcloud-java-storage/src/test/java/com/google/gcloud/storage/StorageImplTest.java]: @aozarov I'll add `testReadAllBytesWithOptions`. 
`reader()` and `writer()` functions build the options map and call the constructors of `BlobReadChannelImpl` and `BlobWriteChannelImpl`. `BlobReadChannel` and `BlobWriteChannel` have no method to get options so to check that they are passed correctly I must either add it or use reflection (which I don't think is a good idea). Can you think of a better way of doing this?

**Comment by mziccard** [comment]
@aozarov Exceptions in `StorageRpc` and `StorageImpl` work in a different way from `DatastoreRpc` and `DatastoreImpl`:

1. `StorageRpc` methods throw `StorageException` (that is the same exception thrown by `Storage`, while `DatastoreRpc` methods have their own `DatastoreRpcException`.
2. In `DatastoreImpl` `RetryHelperException` are caught and translated to `DatastoreException` while `StorageImpl` does not catch them. 

Is there a reason for this difference (in particular the second point)? I think we should catch all `RetryHelperException` in `StorageImpl `and throw the `StorageException` that caused the problem instead.

File: gcloud-java-storage/src/test/java/com/google/gcloud/storage/StorageImplTest.java (hunk 6 above)
Comment (aozarov) [inline gcloud-java-storage/src/test/java/com/google/gcloud/storage/StorageImplTest.java]: Can you verify it by triggeing a read or a write and capturing the options from the mock `storageRpcMock` specific methods (such as `read` or `open`)?

File: gcloud-java-storage/src/test/java/com/google/gcloud/storage/StorageImplTest.java (hunk 8 above)
Comment (aozarov) [inline gcloud-java-storage/src/test/java/com/google/gcloud/storage/StorageImplTest.java]: 1 second is not terrible but 10 seconds is a little to much to this kind of test.
I think having it as a constant is a good idea.

**Comment by aozarov** [comment]
@mziccard Yes, good catch, we should definitely translate `RetryHelperException` to `StorageException`. That was my plan and apparently I never got to do that.  

As for (2), Datastore API and SPI had a cleaner separation. In both cases we should **never** expose SPI in the API layer (which I think we don't) but the Datastore is doing a stricter job in not using API in the SPI layer. The main reason for that was time but honestly I am not sure how important that is and can see other cases where API is used in SPI (e.g. [JNDI](http://docs.oracle.com/cd/E17802_01/products/products/jndi/javadoc/)).

**Comment by mziccard** [comment]
@aozarov I think I took care of all you comments. If you agree I'll leave the exceptions stuff to a separate PR as it require some changes to `StorageImpl` that are out of scope in this one.

**Comment by aozarov** [comment]
@mziccard Few other minor comments, mostly about formatting.

Yes, the exception handler changes (and the tests for it) are beyond the scope of this PR.
Do you want to create an issue for it or should I create one?

**Comment by mziccard** [comment]
@aozarov I swear it's the auto-formatter's fault:) I updated last commit according to your comments. I'll open an issue for the exceptions-related changes.

**Comment by aozarov** [comment]
Thanks!

**Pull request merged**